NDTensors/src/lib/UnallocatedArrays/src/abstractunallocatedarray.jl

```diff
@@ -0,0 +1,52 @@
+using FillArrays: FillArrays, getindex_value
+using NDTensors.SetParameters: set_parameters
+using Adapt: adapt
+
+const UnallocatedArray{ElT,N,AxesT,AllocT} = Union{
+  <:UnallocatedFill{ElT,N,AxesT,AllocT},<:UnallocatedZeros{ElT,N,AxesT,AllocT}
+}
+
+@inline Base.axes(A::UnallocatedArray) = axes(parent(A))
+Base.size(A::UnallocatedArray) = size(parent(A))
+function FillArrays.getindex_value(A::UnallocatedArray)
+  return getindex_value(parent(A))
+end
+
+function Base.complex(A::UnallocatedArray)
+  return set_alloctype(
+    complex(parent(A)), set_parameters(alloctype(A), Position{1}(), complex(eltype(A)))
```

Comment:
Seems like this could be simplified by defining `set_eltype`.

Reply:
For example I could imagine:
```julia
Base.complex(a::UnallocatedArray) = set_eltype(a, complex(eltype(a))
```
and then `set_eltype` being defined using some combination of `set_alloctype` and `set_eltype`.

Reply:
I am work through with this is the definition of `set_eltype`. Typically we have defined `set_eltype` on the type of a structure and it returns the unallocated structure with the element type changed
```set_eltype(structT::Type{<:struct}, elt::Type)
structT = set_parameters(structT, Position{1}(), elt)
...
return structT
end
```
do you think this function is more like `convert` or `adapt`?
I tried to make a `set_eltype` function but in the end had to make one for `Zeros` and one for `Fill` because I would expect to do something like 
```
function set_eltype(T::UnallocatedArray, elt::Type)
  f = parent(T)
  ZT = set_eltype(typeof(f), elt)
  set_alloctype(ZT(getindex_value(f), axes(z)),  set_parameters(alloctype(T), Position{1}(), elt))
end
```
but `Zeros` does not have a constructor`Zeros(zero(elt), (2,3))

Reply:
Is the issue that we have only used `set_eltype` for converting types, not instances?

Reply:
The official Julia suggestion for doing this type of element type conversion is using broadcasting, i.e. `Float32.(a)`. Does that work here? If that works, I would rather use that and not add this definition of `set_eltype` on instances, since it's inconsistent with how we've been using it before where it has only been defined on types. Defining `set_eltype` on instances should be done in another PR if we want to do that.

Reply:
Regarding the broadcasting: This is something I have tried, I wasn't sure at the time if that was standard convention and it is nice to learn that it is suggested by official Julia.  Right now this broadcasting does not work for our `UnallocatedArray` type but it does work for generic `AbstractFill` types, i.e.
```
elt = Float32
z = Zeros{elt}((2, 3))
Z = UnallocatedZeros(z, dev(Matrix{eltype(z)}))
Float64.(z) 
  2×3 Fill{Float64}, with entries equal to 0.0
Float64.(Z)
  2×3 Fill{Float64}, with entries equal to 0.0
 ```
I should be able to figure out how to overwrite this broadcasting for `UnallocatedArrays` When I do, what should I do with the `set_eltype` functions that I have made? My initial thought is to remove them from this PR.

Reply:
Yeah, ideally we would remove the definition of `set_eltype` on instances and just use broadcasting to set the element type.

Reply:
Okay I have fixed the broadcast function for `UnallocatedArrays` and removed the `set_eltype` functions.